### PR TITLE
[WIP] Initial steps towards our own rpm.

### PR DESCRIPTION
**Pull request by jessfraz**
This adds building an rpm to `make all`. It is the first steps towards our own
rpm package for docker.

Next steps would include, getting this in. Then adding the rpm to the release 
script to test.docker.com and get.docker.com. I didn't add this in this PR because
I thought having it first in `make all`, which jenkins runs on every PR, would be
a great first step. Then I can make the changes to the release script and install 
scripts.

@tianon feel free to rip this apart. I am sure the `$DEST/postinst $DEST/prerm $DEST/postrm`
scripts need some TLC, I am not as familiar with building rpms :D 

Common items I tried to move to a dotfile, much like how the others work.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -46,6 +46,7 @@ RUN apt-get update && apt-get install -y \
 	python-pip \
 	python-websocket \
 	reprepro \
+    rpm \
```
Comment (tianon) [inline Dockerfile]: indentation :wink:

Comment (jfrazelle) [inline Dockerfile]: lol I caught it and pushed probably right as you saw this dammit

File: project/make.sh [hunk 2]
```diff
@@ -58,6 +58,8 @@ DEFAULT_BUNDLES=(
 	cover
 	cross
 	tgz
+
+	fedora
```
Comment (tianon) [inline project/make.sh]: So, is this RPM Fedora-specific, or generic for all "RPM distributions" ?  I'd rather see us avoid the mistake we made with the current `.deb` where it's supposed to be generic and that limits us and causes bugs that we're going to have to split it out into multiple `.deb`s to fix (which is going to be painful both for us, and for our users).

File: Dockerfile (hunk 1 above)
Comment (tianon) [inline Dockerfile]: :smiling_imp:

File: project/make/.bundle [hunk 3]
```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+
+# helper scripts for creating package bundles
```
Comment (tianon) [inline project/make/.bundle]: Could we call this file `.packages` or something like that instead?  Each script in `project/make/*` is referred to as a "bundle script", and I don't want to overload that term any more than we already have. :+1:

File: project/make/.bundle [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+PKGVERSION="${VERSION//-/'~'}"
+# if we have a "-dev" suffix or have change in Git, let's make this package version more complex so it works better
+if [[ "$VERSION" == *-dev ]] || [ -n "$(git status --porcelain)" ]; then
+	GIT_UNIX="$(git log -1 --pretty='%at')"
+	GIT_DATE="$(date --date "@$GIT_UNIX" +'%Y%m%d.%H%M%S')"
+	GIT_COMMIT="$(git log -1 --pretty='%h')"
+	GIT_VERSION="git${GIT_DATE}.0.${GIT_COMMIT}"
+	# GIT_VERSION is now something like 'git20150128.112847.0.17e840a'
+	PKGVERSION="$PKGVERSION~$GIT_VERSION"
+fi
```
Comment (tianon) [inline project/make/.bundle]: All of this version munging is actually pretty specific to how Debian/`dpkg` (and thus APT) handles version numbers -- we probably need something different (although likely similar) for RPM-based platforms.

File: project/make/fedora [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	update-rc.d docker defaults > /dev/null || true
+fi
+if [ -n "$2" ]; then
+	_dh_action=restart
+else
+	_dh_action=start
+fi
+service docker $_dh_action 2>/dev/null || true
+
+#DEBHELPER#
```
Comment (tianon) [inline project/make/fedora]: `#DEBHELPER#`? :wink:

We're out of my depth on creating RPMs -- I suggest we find someone who actually knows what they're doing, and get their take on these scripts and how they'll need to change, but it'll help to clarify what our goal is with respect to multiple versions / totally generic, and then maybe ask that same someone if it's a reasonable plan too (since we really, _really_ don't want a repeat of what we're dealing with in our `.deb`s WRT Debian vs Ubuntu).

File: project/make/fedora [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+			--after-install $DEST/postinst \
+			--before-remove $DEST/prerm \
+			--after-remove $DEST/postrm \
+			--architecture "$PACKAGE_ARCHITECTURE" \
+			--prefix / \
+			--depends iptables \
+			--description "$PACKAGE_DESCRIPTION" \
+			--maintainer "$PACKAGE_MAINTAINER" \
+			--conflicts docker \
+			--conflicts docker.io \
```
Comment (tianon) [inline project/make/fedora]: This particular line definitely isn't necessary, but having `docker-io` here might be.

File: project/make/.bundle (hunk 3 above)
Comment (jfrazelle) [inline project/make/.bundle]: ya naming is hard

File: project/make/.bundle (hunk 4 above)
Comment (jfrazelle) [inline project/make/.bundle]: ah ok cool was unsure where the "line" was

File: project/make/fedora (hunk 5 above)
Comment (jfrazelle) [inline project/make/fedora]: ya thats how i knew they needed TLC i decided not to touch maybe we need @vbatts help :)

File: project/make.sh (hunk 2 above)
Comment (jfrazelle) [inline project/make.sh]: ya i was trying to keep the naming in the same fashion but should it be rpm and the ubuntu one we just have to keep as that :(

File: project/make/fedora [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+			--name lxc-docker --version "$PKGVERSION" \
+			--architecture "$PACKAGE_ARCHITECTURE" \
+			--depends lxc-docker-$VERSION \
+			--description "$PACKAGE_DESCRIPTION" \
+			--maintainer "$PACKAGE_MAINTAINER" \
+			--url "$PACKAGE_URL" \
+			--license "$PACKAGE_LICENSE" \
+			--rpm-compression gzip \
+			--epoch 0 \
+			-t rpm
```
Comment (tianon) [inline project/make/fedora]: I don't know if we'll run into the same constraints with `.rpm`s as we did with `.deb`s to where this dummy package was necessary.  Another good thing to get a real RPM maintainer's opinion on. :+1:

File: project/make.sh (hunk 2 above)
Comment (tianon) [inline project/make.sh]: IMO, we should rename "ubuntu", since it's not accurate for what that file
is now, unless we fork it and make a "debian" too, but I'd much rather just
do all that as part of my larger ".deb generation" refactoring that's in
progress (because how we build ".deb" files right now is not ideal, let
alone sufficient for our needs anymore)

Comment (jfrazelle) [inline project/make.sh]: can I do that as another PR that handles all that bc yay we def should

Comment (jfrazelle) [inline project/make.sh]: oh nm you are "in progress" :D

Comment (tianon) [inline project/make.sh]: Oh yeah for sure -- sorry if it sounded like I was trying to say that should be smushed into this one too. :no_mouth:

**Comment by jfrazelle** [comment]
@vbatts @lsm5 halp :smiling_imp:

**Comment by thaJeztah** [comment]
@jfrazelle I think @vbatts isn't available until next week (at least, that's what he [mentioned on IRC](https://botbot.me/freenode/docker-maintainers/2015-02-10/?msg=31696306&page=2))

**Comment by jfrazelle** [comment]
yeah I just am leaving this here for him as a present hahaha

**Comment by thaJeztah** [comment]
You just re-branded RPM to "Red Hat *Present* Manager"! :smile_cat:

**Comment by lsm5** [comment]
@jfrazelle willing to help yes :) ... i'll check this out and get back, btw do you know if debhelper or the like can work directly with fedora/centos rpm spec files?

File: project/make/package-rpm [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	GIT_UNIX="$(git log -1 --pretty='%at')"
+	GIT_DATE="$(date --date "@$GIT_UNIX" +'%Y%m%d.%H%M%S')"
+	GIT_COMMIT="$(git log -1 --pretty='%h')"
+	GIT_VERSION="git${GIT_DATE}.0.${GIT_COMMIT}"
+	# GIT_VERSION is now something like 'git20150128.112847.0.17e840a'
+	PKGVERSION="$PKGVERSION~$GIT_VERSION"
+fi
+
+# probably shouldn't be using dpkg here...
+PACKAGE_ARCHITECTURE="$(dpkg-architecture -qDEB_HOST_ARCH)"
```
Comment (vbatts) [inline project/make/package-rpm]: oh dear...

Something like ` rpm --eval "%{_arch}"` on rpm hosts. Though, since this is an parameter passed to FPM, I'm not sure whether fpm is expecting debian `amd64` or rpm `x86_64` style.

Also, since this code should be agnostic to the host running it (don't call dpkg from rpm systems, and vice versa), I recommend going a cleaner route of a `uname -m` and case statement. Not terribly unlike what we do for exporting arch variables on slackware http://slackware.osuosl.org/slackware64-current/source/a/bash/bash.SlackBuild

**Comment by vbatts** [comment]
sadly, i can't review this working as you intend, right now, because `make all` still fails a test due to the systemd cgroup... so it never makes it to the `package-rpm` step

**Comment by tianon** [comment]
"debhelper" definitely can't work with RPM spec files, no :(

**Comment by jfrazelle** [comment]
noting this will eventually close #8825 so we dont forget

**Comment by jfrazelle** [comment]
i will reopen after we fix https://github.com/docker/docker/issues/10690

File: project/make/package-rpm (hunk 8 above)
Comment (estesp) [inline project/make/package-rpm]: Looks like the author of fpm handles arch translations appropriately (you can pass deb-form arch, and when target is rpm, it will convert properly.  (https://github.com/jordansissel/fpm/blob/master/lib/fpm/package/rpm.rb#L205-L217) I agree that `uname -m` will be best here.
